CDS/js/youtube/youtube-upload-lib.js

```diff
@@ -0,0 +1,117 @@
+            reject(new Error('Cannot upload to youtube: missing a title'));
+        });
+    }
+
+    if (!process.env.description) {
+        return new Promise((fulfill, reject) => {
+            reject(new Error('Cannot upload to youtube: missing a description'));
+        });
+    }
+
+    if (!process.env.category_id) {
+        return new Promise((fulfill, reject) => {
+            reject(new Error('Cannot upload to youtube: missing a category id'));
+        });
+    }
+
+    return dataStore.substituteStrings(connection, [process.env.title, process.env.description, process.env.category_id, process.env.access])
+    .then((substitutedStrings) => {
+        var title, description, category_id, status;
+        [title, description, category_id, status] = substitutedStrings;
+
+        return {
+            snippet: {
+                title: title,
+                description: description,
+                categoryId: category_id
+            },
+            status: { privacyStatus: status ? status : 'private'}
+        }
+    });
+}
+
+function getYoutubeData(connection) {
+    return this.getMetadata(connection)
+    .then((metadata) => {
+        const mediaPath = process.env.cf_media_file;
+
+        if (!process.env.cf_media_file) {
+            throw new Error('Cannot upload to youtube: missing media file path. Make sure that media has been specified in the route');
+        }
+
+        if (process.env.owner_channel) {
+            if (!process.env.owner_account) {
+                throw new Error('Cannot upload to youtube: missing account owner');
+            }
+        }
+
+        return dataStore.substituteStrings(connection, [process.env.cf_media_file, process.env.owner_channel, process.env.owner_account])
+        .then((substitutedStrings) => {
+            var mediaPath, ownerChannel, ownerAccount;
+            [mediaPath, ownerChannel, ownerAccount] = substitutedStrings;
+
+            var youtubeData = {
+                'part': 'snippet,status',
+                'resource': metadata,
+                'media': {body: fs.createReadStream(mediaPath)},
+                'uploadType': 'multipart'
+            };
+
+            if (ownerChannel) {
+                youtubeData.onBehalfOfContentOwner = ownerAccount;
+                youtubeData.onBehalfOfContentOwnerChannel = ownerChannel;
+            }
+
+            return youtubeData;
+        });
+    });
+}
+
+function uploadToYoutube(connection) {
+
+    return youtubeAuth.getAuthClient(connection)
+    .then((oauth2) => {
+
+       var youtubeClient = googleapis.youtube({version: YOUTUBE_API_VERSION, auth: oauth2});
+       return this.getYoutubeData(connection)
+        .then((youtubeData) => {
+
+           return new Promise((fulfill, reject) => {
+                youtubeClient.videos.insert(youtubeData, (err, result) => {
+                    if (err) reject(err);
+                    if (result) {
+                        fulfill(
```

Comment:
I think it is unlikely that the datastore would fail. But if it does, there probably should be something in the logs explaining what happened, that video is on youtube but adding it to the media atom is going to fail. So maybe this promise should be fulfilled with a log, adding the asset to youtube will then fail but it will be obvious why it failed from the logs?